File: app/src/main/java/com/github/premnirmal/ticker/model/StocksProvider.java
package com.github.premnirmal.ticker.model;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.SystemClock;
import android.text.TextUtils;
import android.util.Log;

import com.github.premnirmal.ticker.CrashLogger;
import com.github.premnirmal.ticker.RxBus;
import com.github.premnirmal.ticker.Tools;
import com.github.premnirmal.ticker.events.StockUpdatedEvent;
import com.github.premnirmal.ticker.network.Stock;
import com.github.premnirmal.ticker.network.StocksApi;

import org.joda.time.DateTime;
import org.joda.time.DateTimeZone;
import org.joda.time.format.ISODateTimeFormat;

import java.text.DateFormatSymbols;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;
import java.util.TimeZone;

import javax.inject.Singleton;

import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by premnirmal on 12/21/14.
 */
@Singleton
public class StocksProvider implements IStocksProvider {

    private static final String STOCK_LIST = "STOCK_LIST";
    public static final String SORTED_STOCK_LIST = "SORTED_STOCK_LIST";
    private static final String LAST_FETCHED = "LAST_FETCHED";
    private static final String POSITION_LIST = "POSITION_LIST";

    private static final String DEFAULT_STOCKS = "^SPY,GOOG,AAPL,MSFT,YHOO,TSLA";

    public static final List<String> GOOGLE_SYMBOLS = Arrays.asList(".DJI", ".IXIC");
    public static final List<String> _GOOGLE_SYMBOLS = Arrays.asList("^DJI", "^IXIC");

    private static final Set<String> DEFAULT_SET = new HashSet<String>() {
        {
            add("^SPY");
            add("GOOG");
            add("AAPL");
            add("MSFT");
            add("YHOO");
            add("TSLA");
        }
    };

    private List<String> tickerList;
    private List<Stock> stockList;
    private List<Stock> positionList;
    private String lastFetched;
    private final SharedPreferences preferences;

    private final StocksApi api;
    private final Context context;
    private final RxBus bus;
    private final StocksStorage storage;

    public StocksProvider(StocksApi api, RxBus bus, Context context, SharedPreferences sharedPreferences) {
        this.bus = bus;
        this.api = api;
        this.context = context;
        storage = new StocksStorage(context);
        this.preferences = sharedPreferences;
        final String tickerListVars = preferences.getString(SORTED_STOCK_LIST, DEFAULT_STOCKS);
        tickerList = new ArrayList<>(Arrays.asList(tickerListVars.split(",")));
        final List<String> newTickerList = new ArrayList<>();
        for (String ticker : tickerList) {
            if (ticker != null) {
                newTickerList.add(ticker.replaceAll(".", ""));
            }
        }
        tickerList.removeAll(_GOOGLE_SYMBOLS); // removed google finance because it's causing lots of problems, returning 400s
        if (preferences.contains(STOCK_LIST)) { // for users using older versions
            final Set<String> deprecatedTickerSet = preferences.getStringSet(STOCK_LIST, DEFAULT_SET);
            preferences.edit().remove(STOCK_LIST).apply();
            for (String ticker : deprecatedTickerSet) {
                if (!tickerList.contains(ticker)) {
                    tickerList.add(ticker);
                }
            }
        }

        positionList = Tools.stringToPositions(preferences.getString(POSITION_LIST, ""));

        final String tickerList = Tools.toCommaSeparatedString(this.tickerList);
        preferences.edit().putString(SORTED_STOCK_LIST, tickerList).apply();
        lastFetched = preferences.getString(LAST_FETCHED, null);
        if (lastFetched == null) {
            fetch();
        } else {
            fetchLocal();
        }
    }

    private void fetchLocal() {
        stockList = storage.readSynchronous();
        if (stockList != null) {
            sortStockList();
            sendBroadcast();
            removeGoogleStocks();
        } else {
            fetch();
        }
    }

    private void removeGoogleStocks() {
        if (stockList != null) {
            final Stock dummy1 = new Stock();
            dummy1.symbol = "^DJI";
            final Stock dummy2 = new Stock();
            dummy2.symbol = "^IXIC";
            final Stock dummy3 = new Stock();
            dummy3.symbol = ".DJI";
            final Stock dummy4 = new Stock();
            dummy4.symbol = ".IXIC";
            stockList.remove(dummy1);
            stockList.remove(dummy2);
            stockList.remove(dummy3);
            stockList.remove(dummy4);
        }
    }

    private void save() {
        preferences.edit()
                .remove(STOCK_LIST)
                .putString(POSITION_LIST, Tools.positionsToString(positionList))
                .putString(SORTED_STOCK_LIST, Tools.toCommaSeparatedString(tickerList))
                .putString(LAST_FETCHED, lastFetched)
                .apply();
        removeGoogleStocks();
        storage.save(stockList)
                .subscribe(new Subscriber<Boolean>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        e.printStackTrace();
                    }

                    @Override
                    public void onNext(Boolean success) {
                        if (!success) {
                            Log.e(getClass().getSimpleName(), "Save failed");
                        }
                    }
                });
    }

    @Override
    public void fetch() {
        if (Tools.isNetworkOnline(context)) {
            api.getStocks(tickerList)
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribeOn(Schedulers.io())
                    .subscribe(new Subscriber<List<Stock>>() {
                        @Override
                        public void onCompleted() {
                        }

                        @Override
                        public void onError(Throwable e) {
                            CrashLogger.logException(new RuntimeException("Encountered onError when fetching stocks", e)); // why does this happen?
                            e.printStackTrace();
                            scheduleUpdate(SystemClock.elapsedRealtime() + (60 * 1000)); // 1 minute
                        }

                        @Override
                        public void onNext(List<Stock> stocks) {
                            if (stocks == null || stocks.isEmpty()) {
                                onError(new NullPointerException("stocks == null"));
                            } else {
                                stockList = stocks;
                                lastFetched = api.lastFetched;
                                save();
                                sendBroadcast();
                            }
                        }
                    });
        } else {
            scheduleUpdate(SystemClock.elapsedRealtime() + (5 * 60 * 1000)); // 5 minutes
        }
    }

    private void sendBroadcast() {
        AlarmScheduler.sendBroadcast(context);
        bus.post(new StockUpdatedEvent());
        scheduleUpdate(getMsToNextAlarm());
    }

    /**
     * Takes care of weekends and afterhours
     *
     * @return
     */
    private long getMsToNextAlarm() {
        return AlarmScheduler.msToNextAlarm(context, preferences);
    }

    private void scheduleUpdate(long msToNextAlarm) {
        AlarmScheduler.scheduleUpdate(msToNextAlarm, context, preferences);
    }

    @Override
    public Collection<String> addStock(String ticker) {
        if (tickerList.contains(ticker)) {
            return tickerList;
        }
        tickerList.add(ticker);
        save();
        fetch();
        return tickerList;
    }

    @Override
    public Collection<String> addPosition(String ticker, float shares, float price) {
        Stock position = getStock(ticker);
        if (position == null) {
            position = new Stock();
        } else {

        }
        if(!ticker.contains(ticker)) {
            tickerList.add(ticker);
        }
        position.symbol = ticker;
        position.IsPosition = true;
        position.PositionPrice = price;
        position.PositionShares = shares;
        positionList.remove(position);
        positionList.add(position);
        stockList.remove(position);
        stockList.add(position);
        save();
        fetch();
        return tickerList;
    }

    @Override
    public Collection<String> addStocks(Collection<String> tickers) {
        for (String ticker : tickers) {
            if (!tickerList.contains(ticker)) {
                tickerList.add(ticker);
            }
        }
        save();
        fetch();
        return tickerList;
    }

    @Override
    public Collection<String> removeStock(String ticker) {
        tickerList.remove(ticker);
        final Stock dummy = new Stock();
        dummy.symbol = ticker;
        stockList.remove(dummy);
        positionList.remove(dummy);
        save();
        sendBroadcast();
        return tickerList;
    }

    @Override
    public Collection<Stock> getStocks() {
        if (stockList != null) {
            sortStockList();

            List<Stock> newStockList = new ArrayList<Stock>();
            boolean added = false;
            // Set all positions
            for (Stock stock : stockList) {
                added = false;
                for (Stock pos : positionList) {
                    if (added == false && stock.symbol.equals(pos.symbol)) {
                        stock.IsPosition = true;
                        stock.PositionShares = pos.PositionShares;
                        stock.PositionPrice = pos.PositionPrice;
                        newStockList.add(stock);
                        added = true;
                    }
                }
                if (added == false) {
                    newStockList.add(stock);
                }
            }

            return newStockList;
        }
        return stockList;
    }

    private void sortStockList() {
        if (Tools.autoSortEnabled()) {
            Collections.sort(stockList);
        } else {
            Collections.sort(stockList, new Comparator<Stock>() {
                @Override
                public int compare(Stock lhs, Stock rhs) {
                    return ((Integer) tickerList.indexOf(lhs.symbol))
                            .compareTo(tickerList.indexOf(rhs.symbol));
                }
            });
        }
    }

    @Override
    public Collection<Stock> rearrange(List<String> tickers) {
        tickerList = new ArrayList<>(tickers);
        save();
        sendBroadcast();
        return getStocks();
    }

    @Override
    public Stock getStock(String ticker) {
        final Stock dummy = new Stock();
        dummy.symbol = ticker;
        final int index = stockList.indexOf(dummy);
        if (index >= 0) {
            final Stock stock = stockList.get(index);
            return stock;
        }
        return null;
    }

    @Override
    public List<String> getTickers() {
        return new ArrayList<>(tickerList);
    }

    @Override
    public String lastFetched() {
        final String fetched;
        if (!TextUtils.isEmpty(lastFetched)) {
            final DateTime time = DateTime.parse(lastFetched)
                    .withZone(DateTimeZone.forTimeZone(TimeZone.getDefault()));
            final DateFormatSymbols dfs = DateFormatSymbols.getInstance(Locale.ENGLISH);
            final int fetchedDayOfWeek = time.dayOfWeek().get();
            final int today = DateTime.now().dayOfWeek().get();
            if (today == fetchedDayOfWeek) {
                fetched = time.toString(ISODateTimeFormat.hourMinute());
            } else {
                fetched = dfs.getWeekdays()[fetchedDayOfWeek % 7 + 1] + " "
                        + (time.toString(ISODateTimeFormat.hourMinute()));
            }
        } else {
            fetched = "";
        }
        return fetched;
    }

}

File: app/src/main/kotlin/com/github/premnirmal/ticker/model/StocksProvider.kt
package com.github.premnirmal.ticker.model

import android.content.Context
import android.content.SharedPreferences
import com.github.premnirmal.ticker.AppPreferences
import com.github.premnirmal.ticker.components.AppClock
import com.github.premnirmal.ticker.components.Injector
import com.github.premnirmal.ticker.model.IStocksProvider.FetchState
import com.github.premnirmal.ticker.network.StocksApi
import com.github.premnirmal.ticker.network.data.Holding
import com.github.premnirmal.ticker.network.data.Position
import com.github.premnirmal.ticker.network.data.Quote
import com.github.premnirmal.ticker.repo.StocksStorage
import com.github.premnirmal.ticker.widget.WidgetDataProvider
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Created by premnirmal on 2/28/16.
 */
@Singleton
class StocksProvider : IStocksProvider {

  companion object {

    private const val LAST_FETCHED = "LAST_FETCHED"
    private const val NEXT_FETCH = "NEXT_FETCH"
    private val DEFAULT_STOCKS = arrayOf("^GSPC", "^DJI", "GOOG", "AAPL", "MSFT")
  }

  @Inject lateinit var api: StocksApi
  @Inject lateinit var context: Context
  @Inject lateinit var preferences: SharedPreferences
  @Inject lateinit var appPreferences: AppPreferences
  @Inject lateinit var widgetDataProvider: WidgetDataProvider
  @Inject lateinit var alarmScheduler: AlarmScheduler
  @Inject lateinit var clock: AppClock
  @Inject lateinit var storage: StocksStorage
  @Inject lateinit var coroutineScope: CoroutineScope

  private val exponentialBackoff: ExponentialBackoff

  private val tickerSet: MutableSet<String> = HashSet()
  private val quoteMap: MutableMap<String, Quote> = HashMap()
  private val _fetchState = MutableStateFlow<FetchState>(FetchState.NotFetched)
  private val _nextFetch = MutableStateFlow<Long>(0)
  private val _lastFetched = MutableStateFlow<Long>(0)
  private val _tickers = MutableStateFlow<List<String>>(emptyList())
  private val _portfolio = MutableStateFlow<List<Quote>>(emptyList())

  init {
    Injector.appComponent.inject(this)
    exponentialBackoff = ExponentialBackoff()
    val tickers = storage.readTickers()
    this.tickerSet.addAll(tickers)
    if (this.tickerSet.isEmpty()) {
      this.tickerSet.addAll(DEFAULT_STOCKS)
    }
    _tickers.tryEmit(tickerSet.toList())
    val lastFetched = preferences.getLong(LAST_FETCHED, 0L)
    _lastFetched.tryEmit(lastFetched)
    val nextFetch = preferences.getLong(NEXT_FETCH, 0L)
    _nextFetch.tryEmit(nextFetch)
    coroutineScope.launch {
      alarmScheduler.enqueuePeriodicRefresh(context)
    }
    runBlocking { fetchLocal() }
    if (lastFetched == 0L) {
      coroutineScope.launch {
        fetch().collect()
      }
    } else {
      _fetchState.tryEmit(FetchState.Success(lastFetched))
    }
  }

  private suspend fun fetchLocal() = withContext(Dispatchers.IO) {
      try {
        val quotes = storage.readQuotes()
        synchronized(quoteMap) {
          quotes.forEach { quoteMap[it.symbol] = it }
        }
        _portfolio.emit(quoteMap.values.toList())
      } catch (e: Exception) {
        Timber.w(e)
      }
  }

  private fun saveLastFetched() {
    preferences.edit()
        .putLong(LAST_FETCHED, _lastFetched.value)
        .apply()
  }

  private fun saveTickers() {
    storage.saveTickers(tickerSet)
  }

  private fun scheduleUpdate() {
    scheduleUpdateWithMs(alarmScheduler.msToNextAlarm(_lastFetched.value))
  }

  private fun scheduleUpdateWithMs(
    msToNextAlarm: Long
  ) {
    val updateTime = alarmScheduler.scheduleUpdate(msToNextAlarm, context)
    _nextFetch.tryEmit(updateTime.toInstant().toEpochMilli())
    preferences.edit()
        .putLong(NEXT_FETCH, _nextFetch.value)
        .apply()
    appPreferences.setRefreshing(false)
    widgetDataProvider.broadcastUpdateAllWidgets()
  }

  private fun fetchStockInternal(ticker: String, allowCache: Boolean): Flow<FetchResult<Quote>> =
    flow {
      val quote = if (allowCache) quoteMap[ticker] else null
      quote?.let { emit(FetchResult.success(quote)) } ?: run {
        try {
          emit(api.getStock(ticker))
        } catch (ex: CancellationException) {
          // ignore
        } catch (ex: Exception) {
          Timber.w(ex)
          emit(FetchResult.failure<Quote>(FetchException("Failed to fetch", ex)))
        }
      }
    }.flowOn(Dispatchers.IO)

  /////////////////////
  // public api
  /////////////////////

  override fun hasTicker(ticker: String): Boolean {
    return tickerSet.contains(ticker)
  }

  override fun fetch(): Flow<FetchResult<List<Quote>>> = flow {
    if (tickerSet.isEmpty()) {
      _fetchState.emit(FetchState.Failure(FetchException("No symbols in portfolio")))
      emit(FetchResult.failure<List<Quote>>(FetchException("No symbols in portfolio")))
    } else {
      try {
        appPreferences.setRefreshing(true)
        widgetDataProvider.broadcastUpdateAllWidgets()
        val fr = api.getStocks(tickerSet.toList())
        if (fr.hasError) {
          throw fr.error
        }
        val fetchedStocks = fr.data
        if (fetchedStocks.isEmpty()) {
          emit(FetchResult.failure<List<Quote>>(FetchException("Refresh failed")))
        } else {
          synchronized(tickerSet) {
            tickerSet.addAll(fetchedStocks.map { it.symbol })
          }
          _tickers.emit(tickerSet.toList())
          // clean up existing tickers
          ArrayList(tickerSet).forEach { ticker ->
            if (!widgetDataProvider.containsTicker(ticker)) {
              removeStock(ticker)
            }
          }
          storage.saveQuotes(fetchedStocks)
          fetchLocal()
          _lastFetched.emit(api.lastFetched)
          _fetchState.emit(FetchState.Success(api.lastFetched))
          saveLastFetched()
          exponentialBackoff.reset()
          scheduleUpdate()
          emit(FetchResult.success(fetchedStocks))
        }
      } catch(ex: CancellationException) {
        val backOffTimeMs = exponentialBackoff.getBackoffDurationMs()
        scheduleUpdateWithMs(backOffTimeMs)
      } catch (ex: Throwable) {
        Timber.w(ex)
        _fetchState.emit(FetchState.Failure(ex))
        val backOffTimeMs = exponentialBackoff.getBackoffDurationMs()
        scheduleUpdateWithMs(backOffTimeMs)
        emit(FetchResult.failure<List<Quote>>(FetchException("Failed to fetch", ex)))
      } finally {
        appPreferences.setRefreshing(false)
      }
    }
  }.flowOn(Dispatchers.IO)

  override fun schedule() {
    scheduleUpdate()
    coroutineScope.launch {
      alarmScheduler.enqueuePeriodicRefresh(context, force = true)
    }
  }

  override fun addStock(ticker: String): Collection<String> {
    synchronized(quoteMap) {
      if (!tickerSet.contains(ticker)) {
        tickerSet.add(ticker)
        val quote = Quote()
        quote.symbol = ticker
        quoteMap[ticker] = quote
        saveTickers()
      }
    }
    _tickers.tryEmit(tickerSet.toList())
    _portfolio.tryEmit(quoteMap.values.toList())
    coroutineScope.launch {
      fetchStockInternal(ticker, false).collect { result ->
        if (result.wasSuccessful) {
          val data = result.data
          quoteMap[ticker] = data
          storage.saveQuote(result.data)
          _portfolio.tryEmit(quoteMap.values.toList())
        }
      }
    }
    return tickerSet
  }

  override fun hasPositions(): Boolean = quoteMap.filter { it.value.hasPositions() }.isNotEmpty()

  override fun hasPosition(ticker: String): Boolean = quoteMap[ticker]?.hasPositions() ?: false

  override fun getPosition(ticker: String): Position? = quoteMap[ticker]?.position

  override suspend fun addHolding(
    ticker: String,
    shares: Float,
    price: Float
  ): Holding {
    val quote: Quote?
    var position: Position
    synchronized(quoteMap) {
      quote = quoteMap[ticker]
      position = getPosition(ticker) ?: Position(ticker)
      if (!tickerSet.contains(ticker)) {
        tickerSet.add(ticker)
      }
    }
    _tickers.emit(tickerSet.toList())
    saveTickers()
    val holding = Holding(ticker, shares, price)
    position.add(holding)
    quote?.position = position
    val id = storage.addHolding(holding)
    holding.id = id
    _portfolio.emit(quoteMap.values.toList())
    return holding
  }

  override suspend fun removePosition(
    ticker: String,
    holding: Holding
  ) {
    synchronized(quoteMap) {
      val position = getPosition(ticker)
      val quote = quoteMap[ticker]
      position?.remove(holding)
      quote?.position = position
    }
    storage.removeHolding(ticker, holding)
    _portfolio.emit(quoteMap.values.toList())
  }

  override fun addStocks(symbols: Collection<String>): Collection<String> {
    synchronized(this.tickerSet) {
      val filterNot = symbols.filterNot { this.tickerSet.contains(it) }
      filterNot.forEach { this.tickerSet.add(it) }
      saveTickers()
      if (filterNot.isNotEmpty()) {
        coroutineScope.launch {
          fetch().collect()
        }
      }
    }
    _tickers.tryEmit(tickerSet.toList())
    _portfolio.tryEmit(quoteMap.values.toList())
    return this.tickerSet
  }

  override suspend fun removeStock(ticker: String): Collection<String> {
    synchronized(quoteMap) {
      tickerSet.remove(ticker)
      saveTickers()
      quoteMap.remove(ticker)
    }
    storage.removeQuoteBySymbol(ticker)
    _tickers.emit(tickerSet.toList())
    _portfolio.emit(quoteMap.values.toList())
    return tickerSet
  }

  override suspend fun removeStocks(symbols: Collection<String>) {
    synchronized(quoteMap) {
      symbols.forEach {
        tickerSet.remove(it)
        quoteMap.remove(it)
      }
    }
    storage.removeQuotesBySymbol(symbols.toList())
    _tickers.emit(tickerSet.toList())
    _portfolio.emit(quoteMap.values.toList())
    saveTickers()
  }

  override fun fetchStock(ticker: String): Flow<FetchResult<Quote>> {
    return fetchStockInternal(ticker, true)
  }

  override fun getStock(ticker: String): Quote? = quoteMap[ticker]

  override val tickers: StateFlow<List<String>>
    get() = _tickers

  override val portfolio: StateFlow<List<Quote>>
    get() = _portfolio//quoteMap.filter { widgetDataProvider.containsTicker(it.key) }.map { it.value }

  override fun addPortfolio(portfolio: List<Quote>) {
    synchronized(quoteMap) {
      portfolio.forEach {
        val symbol = it.symbol
        if (!tickerSet.contains(symbol)) tickerSet.add(symbol)
        quoteMap[symbol] = it
      }
    }
    saveTickers()
    widgetDataProvider.updateWidgets(tickerSet.toList())
    coroutineScope.launch {
      storage.saveQuotes(portfolio)
      fetchLocal()
      fetch().collect()
    }
  }

  override val fetchState: StateFlow<FetchState>
    get() = _fetchState

  override val nextFetchMs: StateFlow<Long>
    get() = _nextFetch
}
